### ISSUE TITLE
`function-no-unknown` should check `@container` functions

### ISSUE THREAD

**Issue by Mouvedia** [stylelint/stylelint#7799]
### What minimal example or steps are needed to reproduce the bug?

```css
@container scroll-state() {}
@container style() {}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "function-no-unknown": true
  }
}
```

### How did you run Stylelint?

https://stylelint.io/demo

### Which Stylelint-related dependencies are you using?

.

### What did you expect to happen?

`Unexpected unknown function "scroll-state"`
`Unexpected unknown function "style"`

### What actually happened?

.

### Do you have a proposal to fix the bug?

- `style()`
https://developer.mozilla.org/en-US/docs/Web/CSS/@container#container_style_queries

- `scroll-state()`
chromium flag `CSSScrollStateContainerQueries`
ref https://chromium.googlesource.com/chromium/src/+/aa6104348317fd396bc7965b8adb70f6b8d89ecd%5E%21/#F12

**Comment by ybiquitous** [comment]
@Mouvedia Thanks for the report. I'd like to wait and see how the experimental spec goes for a while.

Even if these functions are added to `css-functions-list`, they will not be detected because `function-no-unknown` scans declarations, not at-rules, at this time.

https://github.com/stylelint/stylelint/blob/a6b9b92e3f7d09d4b044b2eb24b9007ab33230f1/lib/rules/function-no-unknown/index.mjs#L48

So, one more point is that `function-no-unknown` should scan at-rules or not.

**Comment by Mouvedia** [comment]
> I'd like to wait and see how the experimental spec goes for a while.

`style()` is not really experimental at this point.
cf https://caniuse.com/css-container-queries-style

I agree for `scroll-state()` though.

**Comment by ybiquitous** [comment]
In another way, we may add `container-query-no-invalid` like [`media-query-no-invalid`](https://stylelint.io/user-guide/rules/media-query-no-invalid).

**Comment by Mouvedia** [comment]
Even if we were to add a new rule, that may have some intersection with the false positive that `function-no-unknown` should currently have, `@container` should be checked because these are considered functions.

ref
> [Not a style function with space before '('](https://github.com/web-platform-tests/wpt/blob/8b16a577012afd19e6f1de90b4f568c1db0cdd54/css/css-contain/container-queries/at-container-style-serialization.html#L30)

> [Remove the size function syntax from <size-feature>](https://www.w3.org/TR/css-contain-3/#changes-2021-12)

**Comment by ybiquitous** [comment]
I doubt if `function-no-unknown` could support at-rules, not only declarations. To avoid this, adding a new rule like `container-query-no-invalid` sounds reasonable to me.

I wonder if there are other cases of functions in at-rules. 🤔

**Comment by Mouvedia** [comment]
> I wonder if there are other cases of functions in at-rules. 🤔

There's only one more AFAIK: [`selector`](https://developer.mozilla.org/en-US/docs/Web/CSS/@supports#function_syntax).

**Comment by Mouvedia** [comment]
> I wonder if there are other cases of functions in at-rules. 🤔 

There are several more.
see https://developer.mozilla.org/en-US/docs/Web/CSS/@supports#function_syntax

**Comment by ybiquitous** [comment]
Thanks for the info. Since some functions are already supported in at-rules, adding at-rule support to `function-no-unknown` may be possible.

The implementation might be as follows:

```js
// .../function-no-unknown/index.mjs
import valueParser from 'postcss-value-parser';

root.walkAtRules(/^(container|supports)$/i, (atRule) => {
	const valueNodes = valueParser(getAtRuleParams(atRule)).nodes;

	switch (atRule.name.toLowerCase()) {
		case "container":
			// Check unknown functions in `@container` rule...
		case "supports":
			// Check unknown functions in `@supports` rule...
	}
});
```

Are there other points we missed? If not, we could go forward.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for providing the important perspective. As your examples indicate, If we may enhance checks for `@container` and `@supports` in the future, it seems to make sense to implement them as a dedicated rule instead of the more generic `function-no-unknown` rule.

@Mouvedia What do you think about this point?

**Comment by Mouvedia** [comment]
I think both can be pursued.
i.e. it's not incompatible to do both

Since the ones working on the specification and the documentation consider these functions, it should be detected by `function-no-unknown`.
e.g. it's unexpected that we don't have an error for `styles()`, `selectors()` or `font-technology()`
I would be interested to fix it.

In practice, I am pretty sure that we will go a long way to avoid false negatives.
i.e. in the process of determining the parts where a function is allowed we will have a lot a code in common with the upcoming `supports-no-invalid` and `container-no-invalid` rules

tl;dr: the one who will fix these `function-no-unknown` bugs will have to create utils

**Comment by ybiquitous** [comment]
If `function-no-unknown` would scan at-rules too, Stylelint would emit duplicate lint problems caused by both `function-no-unknown` and `container-no-invalid`, right? If so, I worried this would be against the following rule development policy. 🤔 

https://github.com/stylelint/stylelint/blob/cf8479c2424183ff4ff9d6f553606fceda3fb2d1/docs/developer-guide/rules.md?plain=1#L19

**Comment by Mouvedia** [comment]
> singular purpose that doesn't overlap with other rules

I didn't really wanna go into details about the implementation in this issue.
But something like a complicated regular expression—e.g. `/\(([^()]|\(([^()])*\))*\)/g`—could be used to ignore the inside of the parentheses. We would then check the rest.
i.e. it wouldn't be as precise as a `*-no-invalid` rule

It would just check every word which precedes a `(` _at the first level_—i.e. not inside the parentheses.
Since space matters—e.g. `foo()`and `foo ()` are not equivalent—such a basic regex solution would detect unknown functions without overlapping with a much more advanced _non-existing_ yet `*-no-invalid` rule.

**Comment by romainmenke** [comment]
We should avoid regexp and follow this guidance : https://github.com/stylelint/stylelint/blob/cf8479c2424183ff4ff9d6f553606fceda3fb2d1/docs/developer-guide/rules.md?plain=1#L96-L104

But yes, we can definitely have cut off points.

This rule already uses `@csstools/css-parser-algorithms` and the `walk` function used there accepts a third argument for state. This state is passed down to child nodes, but mutations do not bleed to sibling nodes. In other words, it is trivial to keep track of context.

```js
walk(yourAST_Nodes, ({ node, state }) => {
  if (!isFunctionNode(node)) return;

  const name = node.getName();

  if (name === 'calc') state.context = 'math-function';
  if (name === 'supports') state.context = 'supports-condition';
  // ...
}, {
  context: 'root' // or anything that makes sense, maybe `null`?
})
```

You can make these conditions more refined.
For example checking the PostCSS ast node and the specialized parser ast nodes :

```js
if (postcssNode.name === 'import' && name === 'supports') state.context = 'supports-condition';
```

`@csstools/css-parser-algorithms` can also handle any kind of CSS source, it won't throw errors when using it for selectors, at-rule preludes or declaration values.

It won't produced an AST specialized for selectors, but it works well enough to parse `@supports selector(:is(.foo))` without error.

Once the state object indicates that you are inside a `selector` function in a `supports condition` you can return early.

The state can be anything that is easily clone-able.
It can be an array, a string, multiple fields, ...

**Comment by Mouvedia** [comment]
A regex might too brittle or less readable, but maybe it could be more performant?
Anyway my point was, it shouldn't go too far.

I am still of the opinion that not detecting these functions is unexpected.
I won't pick up this issue for a while though.
I have already got a project planned after the refactors.
Unless the rules exist or have at least their own issue proposed and accepted, I don't consider the

> singular purpose that doesn't overlap with other rules

policy a problem.

**Comment by ybiquitous** [comment]
Thanks for the replies. I still lean toward the idea of adding new rules for `@container` and `@supports` instead of adding the at-rules support to `function-no-unknown`.

But as @Mouvedia commented, we can leave this discussion open unless new perspectives come.

**Comment by ybiquitous** [comment]
By the way, I wonder if we should add `@csstools/css-parser-algorithms` and `@csstools/css-tokenizer` to the following recommendation list: 😃 

https://github.com/stylelint/stylelint/blob/cf8479c2424183ff4ff9d6f553606fceda3fb2d1/docs/developer-guide/rules.md?plain=1#L98-L102

**Comment by Mouvedia** [comment]
@jeddy3 does #8159 change something regarding https://github.com/stylelint/stylelint/issues/7799#issuecomment-2190273685?

**Comment by jeddy3** [comment]
Yes, I believe so. It's okay to have overlapping rules. If possible, those rules should have secondary options so that they can be used together, where appropriate.
